Separate `DefId`s for variants and their constructors

davidtwco opened the pull request:
Part of #44109. Split off from #59376. See [Zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/rfc-2008/near/132663140) for previous discussion.

r? @petrochenkov

bors commented:
:umbrella: The latest upstream changes (presumably #59096) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
I reviewed the PR and had comments, but many of them required some experimentation, so I basically started tweaking the code on top of this PR.
I'll hopefully continue and finish making these changes tomorrow.

petrochenkov commented:
Any use of the is_struct/is_tuple/is_unit methods are suspicious (as you could notice, `!is_struct` was usually used as `has_ctor`).
(Exhaustive) matching should generally be used instead.
Some code in derive machinery still needs this, but that's because it's *very* legacy.

Getting from `VariantDef` to its `AdtDef` is an uncommon operation, so it's better to remove `parent_did` from it + the optional `DefId` introduced some complexity/confusion that wasn't useful in practice.

petrochenkov commented:
>I'll hopefully continue and finish making these changes tomorrow.

Done.

One remaining issue is that `Node::Ctor` should not contain `CtorOf` if possible.
`CtorOf` can be determined by looking at the node's parent in few places where it makes difference.
I tried to do that in https://github.com/petrochenkov/rust/commit/63edbb55374ecca5b6f9247ef82802ab3c23196f, but it turned out that variant ctors and variants themselves are not always connected by the parent-child relation (variant is not a variant ctor's parent), and I didn't investigate how to fix that.

After the `Node::Ctor` is done, it will be better for `CtorOf` to live in `def.rs` together with `CtorKind`.
(And one more thing that annoys me unproportionally is that in `Ctor(hir::CtorOf, DefId, CtorKind)` the "secondary" thing (`CtorOf`) is placed before the "primary" thing (`DefId`), I'd personally change it to `Ctor(DefId, CtorOf, CtorKind)` or something.)

davidtwco commented:
Thanks for all the changes.

> One remaining issue is that Node::Ctor should not contain CtorOf if possible.
CtorOf can be determined by looking at the node's parent in few places where it makes difference.
I tried to do that in petrochenkov@63edbb5, but it turned out that variant ctors and variants themselves are not always connected by the parent-child relation (variant is not a variant ctor's parent), and I didn't investigate how to fix that.

I've fixed this. Your patch was pretty much there, but (the misleadingly named) `get_parent` function was returning a `Node::Item` as it finds the next parent that is an item, so I used `get_parent_node` and then it just worked.

> After the Node::Ctor is done, it will be better for CtorOf to live in def.rs together with CtorKind.
(And one more thing that annoys me unproportionally is that in Ctor(hir::CtorOf, DefId, CtorKind) the "secondary" thing (CtorOf) is placed before the "primary" thing (DefId), I'd personally change it to Ctor(DefId, CtorOf, CtorKind) or something.)

Fixed both of these.

petrochenkov commented:
Thanks.
[3.5 years in the making!](https://github.com/rust-lang/rust/pull/28816)
@bors r+

bors commented:
:pushpin: Commit 23cae1d3f06ccb339f2f780e63b6d6d5c1c6a9da has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 23cae1d3f06ccb339f2f780e63b6d6d5c1c6a9da -->

petrochenkov commented:
@bors p=1
(The changes are somewhat conflict-prone and unblock other work.)

bors commented:
:hourglass: Testing commit 23cae1d3f06ccb339f2f780e63b6d6d5c1c6a9da with merge 3752b3d3a56bf3eabb588b7d595cd1f8accc0286...
<!-- homu: {"type":"BuildStarted","head_sha":"23cae1d3f06ccb339f2f780e63b6d6d5c1c6a9da","merge_sha":"3752b3d3a56bf3eabb588b7d595cd1f8accc0286"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/105596734), [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/23321346)
Approved by: petrochenkov
Pushing 3752b3d3a56bf3eabb588b7d595cd1f8accc0286 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang/rust/builds/23321346","checks-travis":"https://travis-ci.com/rust-lang/rust/builds/105596734"},"merge_sha":"3752b3d3a56bf3eabb588b7d595cd1f8accc0286"} -->

RalfJung commented on src/librustc/ty/mod.rs:
```diff
@@ -1801,55 +1807,73 @@ bitflags! {
     pub ctor_kind: CtorKind,
+    /// Flags of the variant (e.g. is field list non-exhaustive)?
     flags: VariantFlags,
+    /// Recovered?
```
I think this is the most redundant doc-comment I have ever seen.^^ It explains literally nothing...

If you have a one-sentence description of what this field does, I can add it in https://github.com/rust-lang/rust/pull/63346. Or maybe it is easier for you to make a PR yourself. :)

davidtwco replied:
I agree that this comment isn't particularly useful. This PR was a while ago, I assume I just wanted to add some comment on each field as I was changing the type - I'm afraid I don't know what the `recovered` field is used for.
